### PR TITLE
go/libraries/doltcore/remotestorage,go/store/datas/pull: Implement a pipelined chunk fetcher.

### PR DESCRIPTION
**Pull request by reltuk**

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>1c6a4cf</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>1c6a4cf</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d900819</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d900819</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Review by max-hoffman — APPROVED** [review]
LGTM, bit more concerned about the hand rolled primitives than the orchestration. Like realistically I'd need to do a bunch of testing to find whether the message queue has any edge cases where we'd drop a req or something. Same thing with the state machine, it all checks out afaict but it seems like there's a lot of potential failure cases where we don't restart a streaming connection or something. But it seems pretty unlikely that we would return successfully from a fetch without detecting missed chunks? So then worst-case is maybe a repo that has a fetch bug that we need to patch before it can be cloned, which is fine?

File: go/libraries/doltcore/remotestorage/internal/ranges/ranges.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+// |intern| will deduplicate strings that are stored in the |ranges.Tree|, so
+// that all equal values share the same heap memory. The context is that URLs
+// stored in the |Tree| can be very long, since they can be pre-signed S3 URLs,
+// for example. And in general a Tree will have a large number of |GetRange|
+// entries, that contain the same |Url|.
+func (t *Tree) intern(s string) string {
+	t.t.AscendGreaterOrEqual(&GetRange{Url: s}, func(gr *GetRange) bool {
+		if gr.Url == s {
+			s = gr.Url
```
[the remark this answers is not in the document]
Comment (reltuk) [reply go/libraries/doltcore/remotestorage/internal/ranges/ranges.go:137]: Hmm...how can I make it more clear? The intent here is that there are `*GetRange` instances stored in the tree, and those instances might have a `Url` matching the incoming `Url`. Essentially, we want to take the existing `StringHeader` of an existing entry if it exists, because that will share backing storage with something that's already in the tree...so, we're assigning `s` here, since it's the thing we're going to return down below.

[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/remotestorage/internal/ranges/ranges.go:137]: yeah ig that makes sense, forward-looking sharing vs backward

File: go/libraries/doltcore/remotestorage/internal/reliable/http.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+				if err != nil {
+					cCause(err)
+				}
+			}()
+
+			httpReq = httpReq.WithContext(ctx)
+
+			req.Stats.RecordDownloadAttemptStart(retry, offset-origOffset, req.Length)
+			start := time.Now()
+			tc.SetTimeout(ctx, req.RespHeadersTimeout)
```
[the remark this answers is not in the document]
Comment (reltuk) [reply go/libraries/doltcore/remotestorage/internal/reliable/http.go:157]: Hmm....maybe I will add a comment. The `ctx` will persist for the duration of the request, so if it ever becomes canceled, the request will fail, including the read of the Body. Here, we want to impose a timeout around the RTT for the initial request headers, but the throughput enforcer will enforce a thing like a timeout for the read of the body. So `context.WithTimeout` doesn't quite work for us AFAIK.

File: go/libraries/doltcore/remotestorage/chunk_fetcher.go [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+
+		select {
+		case hs, ok := <-reqCh:
+			if !ok {
+				reqCh = nil
+				break
+			}
+			for h := range hs {
+				h := h
+				addrs = append(addrs, h[:])
```
[the remark this answers is not in the document]
Comment (reltuk) [reply go/libraries/doltcore/remotestorage/chunk_fetcher.go:199]: Hmm, I'm not sure I understand. You mean why do we batch them up separate from the incoming `HashSet`?

The reason is a few fold, AFAIK:

1. The remote server is going to impose limits on incoming request size, and it wants incoming request messages to be below a certain threshold.
2. The remote server is going to respond to incoming requests serially, and requests are responded to in aggregate / fully – a response message comes back with all the resolved locations, and that response body has to be fully read and processed before any locations can be pulled out of it, etc.
3. The remote server is essentially running `HasMany()` or `getReadReqs` or whatever on the table file indexes of the loaded Dolt repository. The method is more efficient with batches of addresses than for single shot lookups.
4. The response payload itself is more efficient for multi-lookups where the locations share the same table file. In that case, we only have to transit the signed URL for a given table file once. Whereas the response payloads are stateless across the connection, and so the smaller the batches, the more times the signed URLs are transitted.

The end result is that you want to batch up chunk addresses, but you can't / don't want your batches to be too big.

Does that answer the question?

In reality, some of these constraints are self-imposed and could be approached in a different way, but I think you would still want batching here...

File: go/libraries/doltcore/remotestorage/chunk_fetcher.go [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+			BackOffF:           grpcBackOff,
+			ReadRequestTimeout: 15 * time.Second,
+			DeliverRespTimeout: 15 * time.Second,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
```
[the remark this answers is not in the document]
Comment (reltuk) [reply go/libraries/doltcore/remotestorage/chunk_fetcher.go:236]: Hmm...I think in general I prefer to make a new one? Like, if the function itself just has a contract that it's going to run to completion and return an error, then the caller doesn't know or care whether it is using an errgroup to do its job or not.

If the function has a weirder contract that involves explicit concurrency that's somehow controlled or provided by the caller, then things can get more strange...

File: go/libraries/doltcore/remotestorage/chunk_fetcher.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/remotestorage/chunk_fetcher.go:199]: Oh I see, I thought batching was working in the other direction for some reason -- accumulating hash sets, which the setup doesn't seem organized for. If we're sending out smaller batches, decoupling the receive/send makes sense

File: go/libraries/doltcore/remotestorage/internal/ranges/ranges.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (reltuk) [reply go/libraries/doltcore/remotestorage/internal/ranges/ranges.go:137]: Yeah, the thing about forward sharing is that there will be an arbitrary, possibly very large, number of entries already in the tree with `Url == s`

File: go/libraries/doltcore/remotestorage/internal/reliable/chan.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+	// Used by clients to request to peek at the Front of the current
+	// |buff|.
+	front chan frontReq[T]
+
+	// Used by clients to read the current value of |c.output|.
+	outputCh chan chan T
+}
+
+type frontReq[T any] struct {
+	resCh chan T
```
[the remark this answers is not in the document]
Comment (reltuk) [reply go/libraries/doltcore/remotestorage/internal/reliable/chan.go:81]: 🤣

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>b449280</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>b449280</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/remotestorage/chunk_fetcher.go [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+func (f *ChunkFetcher) Close() error {
+	defer StatsFlusher(f.stats)
+	close(f.abortCh)
+	return f.eg.Wait()
+}
+
+// Reads HashSets from reqCh and batches all the received addresses
+// into |GetDownloadLocsRequest| messages with up to |batchSize| chunk hashes
+// in them. It delivers the batched messages to |resCh|.
+func fetcherHashSetToGetDlLocsReqsThread(ctx context.Context, reqCh chan hash.HashSet, abortCh chan struct{}, resCh chan *remotesapi.GetDownloadLocsRequest, batchSize int, repoPath string, idFunc func() (*remotesapi.RepoId, string)) error {
```
Comment (nicktobey) [inline go/libraries/doltcore/remotestorage/chunk_fetcher.go:156]: The state machine in this function is pretty hard to follow. I'm not sure I understand it.

So `thisResCh` is assigned to `resCh` the first time there are any elements in `addr`. Prior to that, it's `nil.` Writing to a `nil` channel blocks forever. Is the idea that the select statement will never choose the `thisResCh <- thisRes` case until we assign `thisResCh` a non-nil value? Are we depending on the properties of writing to a `nil` channel for correctness here? Is this a good practice?

Maybe this is the idiomatic way to do this. But as someone less versed in go channels, the intended execution flow is very unclear.
